src/java.base/share/classes/java/lang/Math.java

```diff
@@ -1346,16 +1347,278 @@ public static long negateExact(long a) {
+            throw newArithmeticExceptionInexact(value);
+        }
+        return result;
+    }
+
+    /**
+     * Returns the value of the {@code long} argument,
+     * throwing an exception if the conversion is inexact.
+     * The method returns iff the argument and the result
+     * are mathematically equal.
+     *
+     * <p>Special case:
+     * <ul>
+     * <li>If the argument is {@code 0L}, the method returns {@code 0.0}.
+     * </ul>
+     *
+     * @param value the {@code long} value
+     * @return the argument as a {@code double}
+     * @throws ArithmeticException if the conversion is inexact
+     * @since 19
+     */
+    @ForceInline
+    public static double toDoubleExact(long value) {
+        /*
+         * The left-hand side of the test below correctly discriminates
+         * inexact conversions, *except* for Long.MAX_VALUE.
+         * Long.MAX_VALUE cannot be converted exactly,
+         * and yet the left-hand side evaluates to false.
+         */
+        double result = (double) value;
+        if ((long) result != value || value == Long.MAX_VALUE) {
+            throw newArithmeticExceptionOverflow(value);
+        }
+        return result;
+    }
+
+    /**
+     * Returns the value of the {@code double} argument,
+     * throwing an exception if the conversion is inexact.
+     * The method returns iff the argument and the result
+     * are mathematically equal.
+     *
+     * <p>Special cases:
+     * <ul>
+     * <li>If the argument is {@link Double#NEGATIVE_INFINITY},
+     * {@link Double#POSITIVE_INFINITY} or {@link Double#NaN},
+     * the method throws.
+     * <li>If the argument is {@code -0.0} or {@code 0.0},
+     * the method returns {@code 0L}.
+     * </ul>
+     *
+     * @param value the {@code double} value
+     * @return the argument as a {@code long}
+     * @throws ArithmeticException if the conversion is inexact
+     * @see Math#rint(double)
+     * @see Math#round(double)
+     * @since 19
+     */
+    @ForceInline
+    public static long toLongExact(double value) {
+        /*
+         * The left-hand side of the test below correctly discriminates
+         * inexact conversions, *except* for 0x1p63.
+         * 0x1p63 cannot be converted exactly,
+         * and yet the left-hand side evaluates to false.
+         */
+        long result = (long) value;
+        if ((double) result != value || value == 0x1p63) {
+            throw newArithmeticExceptionInexact(value);
+        }
+        return result;
+    }
+
+    /**
+     * Returns the value of the {@code long} argument,
+     * throwing an exception if the value overflows the range
+     * [0, 2<sup>{@link Integer#SIZE}</sup>) an unsigned int.
+     *
+     * @param value the {@code long} value
+     * @return the argument as an unsigned int
+     * @throws ArithmeticException if the argument overflows an unsigned int
+     * @since 19
+     */
+    @ForceInline
+    public static long toUnsignedIntExact(long value) {
```

Comment:
Existing methods like Integer.parseUnsignedInt interpret the negative int values as positive values larger than MAX_INT. So if an int is not going to be returned here, I suggest a name like "toUnsignedIntRangeExact".